rafx-framework/src/visibility/view_frustum_object.rs

```diff
@@ -0,0 +1,239 @@
+use crate::RafxResult;
+use crossbeam_channel::Sender;
+use glam::Vec3;
+use parking_lot::{RwLock, RwLockReadGuard};
+use rafx_api::RafxError;
+use rafx_visibility::{
+    AsyncCommand, Projection, ViewFrustumHandle, VisibilityQuery, VisibilityWorldArc, ZoneHandle,
+};
+use slotmap::SlotMap;
+use std::sync::Arc;
+
+#[derive(Clone)]
+pub struct ViewFrustumObjectHandle {
+    inner: Arc<RemoveViewFrustumWhenDropped>,
+}
+
+impl ViewFrustumObjectHandle {
+    pub fn new(
+        static_view_frustum: Option<ViewFrustumObjectId>,
+        dynamic_view_frustum: Option<ViewFrustumObjectId>,
+        storage: SlotMapArc<ViewFrustumObjectId, ViewFrustumObject>,
+        visibility_world: VisibilityWorldArc,
+    ) -> Self {
+        Self {
+            inner: Arc::new(RemoveViewFrustumWhenDropped {
+                static_view_frustum,
+                dynamic_view_frustum,
+                storage,
+                visibility_world,
+                visibility_query: Default::default(),
+            }),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(super) fn set_zone(
+        &self,
+        static_zone: Option<ZoneHandle>,
+        dynamic_zone: Option<ZoneHandle>,
+    ) -> &Self {
+        let storage = self.inner.storage.read();
+
+        if let Some(static_view_frustum) =
+            self.view_frustum(&self.inner.static_view_frustum, &storage)
+        {
+            static_view_frustum.set_zone(static_zone);
+        }
+
+        if let Some(dynamic_view_frustum) =
+            self.view_frustum(&self.inner.dynamic_view_frustum, &storage)
+        {
+            dynamic_view_frustum.set_zone(dynamic_zone);
+        }
+
+        self
+    }
+
+    pub fn set_transform(
+        &self,
+        eye: Vec3,
+        look_at: Vec3,
+        up: Vec3,
+    ) -> &Self {
+        let storage = self.inner.storage.read();
+
+        if let Some(static_view_frustum) =
+            self.view_frustum(&self.inner.static_view_frustum, &storage)
+        {
+            static_view_frustum.set_transform(eye, look_at, up);
+        }
+
+        if let Some(dynamic_view_frustum) =
+            self.view_frustum(&self.inner.dynamic_view_frustum, &storage)
+        {
+            dynamic_view_frustum.set_transform(eye, look_at, up);
+        }
+
+        self
+    }
+
+    pub fn set_projection(
+        &self,
+        projection: &Projection,
+    ) -> &Self {
+        let storage = self.inner.storage.read();
+
+        if let Some(static_view_frustum) =
+            self.view_frustum(&self.inner.static_view_frustum, &storage)
+        {
+            static_view_frustum.set_projection(projection);
+        }
+
+        if let Some(dynamic_view_frustum) =
+            self.view_frustum(&self.inner.dynamic_view_frustum, &storage)
+        {
+            dynamic_view_frustum.set_projection(projection);
+        }
+
+        self
+    }
+
+    pub fn query_visibility(&mut self) -> RafxResult<RwLockReadGuard<VisibilityQuery>> {
```

Comment:
returning a lock feels like a deadlock waiting to happen (but I don't have a useful alternative to suggest atm :D )

Reply:
Yeah, I'm not happy about it either. We'll be able to find a better API as we use it more in examples and continue bringing in functionality.